Add docs for using custom images with Beaker Interactive. (#24)
beaker/docs: docs/start/interactive.md

Your home directory is mounted into the container so anything you write to `~` will persist
between sessions on that node. NFS is also available in the `/net` directory.

### Reserving GPUs

By default, sessions aren't assigned a GPU.

> If you need GPUs, use the `--gpus <count>` flag when creating a session e.g.:

Reviewer: Nit: Should collapse paragraph since these lines are interrelated.

Author: Good catch, done.

If you need to use a different base image, pass the `--image` flag when creating a session.

Beaker Interactive supports both [Docker](https://docker.com) and [Beaker](../concept/images) 

> images. To use a docker image, prefix the image name with `docker://`. To use a

Reviewer: Docker is a trademarked brand name.  We need to capitalize it.
```suggestion
images. To use a Docker image, prefix the image name with `docker://`. To use a
```

Author: Gotcha, can do! I included this in a batch of changes I just pushed up.

Beaker image use the `beaker://` prefix instead.

For example, this command uses the AllenNLP base image and runs the `test-install` command:

```
beaker session create --image docker://allennlp/allennlp -- allennlp test-install
```

> Some features, like user mapping, will not work with all images. For instance, In the AllenNLP 

Reviewer 2: What images will they work on?  Maybe mention that they will work on images that use XYZ as their base image.

Author: I don't know, I didn't write this. I'd have to ask @aaasen or @ckarenz to clarify.

I think this might have to do with the fact that some images might not have bash or even a shell at all -- images can be super slimmed down and accordingly not include the things that regular linux machines do to facility multi-tenancy.

Reviewer: It will only work on the one default image we provide (`allenai/cuda11.2-ubuntu20.04`).  This should be a short-term situation.

Author: Gotcha. I updated the docs.

image, you will see a prompt like `I have no name!@fd82c7800efa:~$`. Please contact the 
Beaker team if you need help setting up a custom environment for your interactive sessions.

### Image Caching

By default Beaker Interactive will use a local copy of an image if one exists. For example, if 

> you've ran `docker pull allennlp/allennlp` before, then starting a session using that image won't

Reviewer: Grammer: `you've ran` -> `you've run`

Author: Thanks! I included this in some changes I just pushed.

pull the latest version to the host:

```
beaker session create --image docker://allennlp/allennlp
```

You can change this behavior using the `--pull` option. The option can be 
one of `always`, `missing` or `never`, and defaults to `missing`. If the value is `always`,
the latest version is always retrieved from the source:

```
beaker session create --image docker://allennlp/allennlp --pull always
```

If the value is `missing` the image will only be pulled if it doesn't already exist locally. 

> Setting the value to `never` means that the image won't ever be pulled, which means you'll see 

Reviewer: ```suggestion
Setting the value to `never` means the image won't ever be pulled, which means you'll see 
```

Author: 👏  Fixed, thanks!

an error if it doesn't exist in the local cache:

```
beaker session create --image docker://php --pull never
Starting session 01F51DJY6PQEWFZRVVDQ7072A2... (Press Ctrl+C to cancel)
Verifying image (docker://php)...

> Error: Error: No such image: php

Reviewer: This stutter is a bug.  Let's change the doc text to the fixed version and I'll put out a change today.
```suggestion
Error: No such image: php
```

Author: Sounds good!

```

### Building Custom Images

At some point you might need to install additional software in your interactive session. You can 

> do this without existing the session, but the changes won't persist after the session 

Author: Fixed.

> stops. To persist those changes you'll need to build a custom image. For the purpose of example 

Reviewer: ```suggestion
stops. To persist those changes you'll need to build a custom image. For example 
```

Author: 👍  Thx!

let's assume we want to create an image with  [jq](https://stedolan.github.io/jq/) installed.

Start by creating a `Dockerfile`:

```
touch Dockerfile
```

If you're unfamiliar with the `Dockerfile` syntax, you can read more about it [here](
https://docs.docker.com/engine/reference/builder/).

Next chose a base image to start from. If you're unsure of what to use it's probably
good to use the [allenai/base](https://hub.docker.com/r/allenai/base/tags) image:

```Dockerfile
FROM allenai/base:cuda11.2-ubuntu20.04
```

Now we'll add a `RUN` entry to the `Dockerfile` that installs `jq`:

```Dockerfile

> RUN wget -O /usr/bin/jq https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64 && \

Reviewer 2: Why run `wget` instead of `apt-get install`?  The former looks a bit scary!  How would someone come up with this complex syntax?

Author: I just followed `jq`'s documentation :).

I'll see if `apt-get` works without repository munging this afternoon and use that instead, as you're right, it's less surprising.

Author: Fixed, thanks!

	chmod 751 /usr/bin/jq
```

After that's done we can build an image using the `docker build` command. The resulting image
will only exist in our local cache: 

```
docker build -t allenai/base:cuda11.2-ubuntu20.04-jq1.6 .
```

Now you can start a session using that image like so:

```
beaker session create --image docker://allenai/base:cuda11.2-ubuntu20.04-jq1.6
```

...and you should now be able to use `jq` like so:

```
[BEAKER] sams ~ $ echo '{ "jq": "is pretty great!" }' | jq
{
  "jq": "is pretty great!"
}
```

At this point the image **only** exists in your local cache. The local image cache is periodically 
flushed, so you'll need to push the image to a remote repository if you'd like to use 

> it again later or share it with other folks at AI2.

Reviewer: ```suggestion
it again later or share it with others.
```

Author: 👍

You can push your image to either [Docker Hub](https://hub.docker.com) or 
[Beaker Images](https://beaker.org/images). If you're unsure of which to use, use Beaker. Beaker

> images are easier to use, can be used in [Beaker Experiments](../concept/experiments) and 

Reviewer 2: Can you be more specific than "easier to use"?  It seems like the main advantage is the fine-grained access control.

Author: Yea, I suppose that's right. I figured the fact that there's a suite of commands related to `beaker images` means we afford some conveniences -- but I don't understand the full capabilities of beaker yet, so it's possible restricting access is the only motivator.

@ckarenz, what do you think?

Reviewer: I agree. Easier to use is debatable.  The public docker registry is pretty simple.  Beaker images only exist because we need access control.

Author: Great, amended. Thx both!

support fine-grained access control.

To push the image to Beaker Images, use the `beaker image create` command. You'll need to execute
this command on the host, not in the interactive session -- which you can exit by either

> typing `exit` or pressing `Ctrl^D`.

Reviewer: I don't think we support the detach interrupt `Ctrl+D`.  That's used to detach from a running container without stopping it, something we're not sure we want to support.

Author: It appears to work, and the session is closed:

```
sams@beaker-server1.corp ~ $ beaker session create
Starting session 01F51PK8RVGM7PVJBDZKVSBY21... (Press Ctrl+C to cancel)
Verifying image (docker://allenai/base:cuda11.2-ubuntu20.04)...

Reserved 26.25 CPUs, 121659562598 B memory
[BEAKER] sams ~ $ logout
```

I see `logout` after hitting `Ctrl+D`. Maybe `docker` is handling this for us?

Reviewer: I'm a little surprised, but that should be OK.  Is the session left running when you log out?

Author: Nope:

```
sams@beaker-server1.corp ~ $ beaker session list
ID  NAME  AUTHOR  STATUS  SCHEDULED  DURATION  GPUS
```

Author: Now, that said, if it's non-standard behavior that might not work in other containers, we should probably remove it.
So I'm going to remove that alternative. Using `exit` seems sufficient.

```
beaker image create allenai/base:cuda11.2-ubuntu20.04-jq1.6 \
    --description "allenai/base with jq too!"
Pushing allenai/base:cuda11.2-ubuntu20.04-jq1.6 as im_j1cp9vbscnz8 ...
```

Once the image is pushed, you can start another interactive session using the image id
in the output of the the create command:

```

> beaker session create --image beaker://im_j1cp9vbscnz8

Reviewer 2: Do we support using names to refer to images rather than IDs?  We've generally been trying to get away from ids, as they are unintelligible to a human user and rather unfriendly.  If we support names, I think we should use them in this example.

Reviewer 2: Yes, this worked fine.  I suggest naming your image something like `base-with-jq` for this example.

```
beaker session create --image beaker://mytestimage
```

Author: Yea, I prefer names too -- but couldn't get them to work. I'll play around this afternoon to figure out what I'm doing wrong (and update the docs).

Author: Ah, so, I was confused as I assumed the original tag is something that could be used as the name. It looks like I need to pass the `--name` argument to `beaker image create`.

That's probably in the documentation for images that I haven't read in full yet.  I'll update these docs to use names throughout once I get back from lunch.

Reviewer: It's a little counter-intuitive.  IMO we should require names images and datasets.

Author: Fixed, thanks.
